onmt/Trainer.py

```diff
@@ -0,0 +1,166 @@
+import torch
+import onmt
+import onmt.modules
+
+
+class Trainer(object):
+    def __init__(self, model, train_data, valid_data, fields, optim,
+                 batch_size, gpuid, copy_attn, copy_attn_force,
+                 truncated_decoder, max_generator_batches):
+        # Basic attributes.
+        self.model = model
+        self.train_data = train_data
+        self.valid_data = valid_data
+        self.fields = fields
+        self.optim = optim
+        self.batch_size = batch_size
+        self.gpuid = gpuid
+        self.copy_attn = copy_attn
+        self.truncated_decoder = truncated_decoder
+        self.max_generator_batches = max_generator_batches
+
+        # Define criterion.
+        padding_idx = fields['tgt'].vocab.stoi[onmt.IO.PAD_WORD]
+        if not copy_attn:
+            self.criterion = onmt.Loss.nmt_criterion(
+                len(fields['tgt'].vocab), gpuid, padding_idx)
+        else:
+            self.criterion = onmt.modules.CopyCriterion(
+                len(fields['tgt'].vocab), copy_attn_force, padding_idx)
+
+        # Create a train data iterator.
```

Comment:
Actually, I think these should be created on the fly each epoch. Train and valid should take train_data and valid_data as arguments. Let's users implement things like curriculum